### ISSUE TITLE
Make extending reused config values easier

### ISSUE THREAD

**Issue by dcastil** [dcastil/tailwind-merge#54]
If someone extends the `spacing` scale in the Tailwind config with values not supported by tailwind-merge, they need to extend the class groups `inset`, `inset-x`, `inset-y`, `top`, `right`, `bottom`, `left`, `gap`, `gap-x`, `gap-y`, `p`, `px`, …

There should be an easier way to extend these scales. It shouldn't just work for spacing, but also border-radius, color, etc.

_Inspired by @nenostra in https://github.com/dcastil/tailwind-merge/discussions/48#discussioncomment-1579038_

**Comment by dcastil** [comment]
Related:

- [Tailwind config reference](https://tailwindcss.com/docs/theme#configuration-reference)
- [Tailwind core plugins reference](https://tailwindcss.com/docs/configuration#core-plugins)

**Comment by nenostra** [comment]
@dcastil Thank you for this! 
Btw, I didn't mention how I ended up solving this, this is what I am doing:

```
import { extendTailwindMerge } from 'tailwind-merge';
import { spacing } from '../config/theme';

const spacingKeys = Object.keys(spacing);

export const twMerge = extendTailwindMerge({
  classGroups: {
    gap: [{ gap: spacingKeys }],
    'gap-x': [{ 'gap-x': spacingKeys }],
    'gap-y': [{ 'gap-y': spacingKeys }],
    p: [{ p: spacingKeys }],
    px: [{ px: spacingKeys }],
    py: [{ py: spacingKeys }],
    pt: [{ pt: spacingKeys }],
    pr: [{ pr: spacingKeys }],
    pb: [{ pb: spacingKeys }],
    pl: [{ pl: spacingKeys }],
    m: [{ m: spacingKeys }],
    mx: [{ mx: spacingKeys }],
    my: [{ my: spacingKeys }],
    mt: [{ mt: spacingKeys }],
    mr: [{ mr: spacingKeys }],
    mb: [{ mb: spacingKeys }],
    ml: [{ ml: spacingKeys }],
    ...
  }
})
```

Then I use my custom `twMerge` through my app. Maybe a simple way to support this (I don't have enough context, this is just what comes on top of my head :P) would be to support `spacing` in the classGroups, and apply those keys to all the spacing css properties.

**Comment by dcastil** [comment]
Okay I thought about this a little bit and have some possible solutions in mind.

# 1. Add theme scales to config

We could add a new `theme` key to the config  which could be used like this:

```ts
extendTailwindMerge({
    theme: {
        spacing: ['0', '2xs', 'xs', '2sm', 'sm', …],
        colors: …,
        somethingElse: …
    }
})
```

Within the default config of tailwind-merge we could then target those scales from the theme object. The easiest implementation of this would be with string keys like `theme.spacing` and then search for strings which start with `theme.` to identify the "theme scale expansion" strings.

```ts
const defaultConfig = {
    classGroups: {
        m: [{ m: ['auto', 'theme.spacing'] }],
        'text-color': [{ text: ['theme.colors'] }],
        …
    },
    …
}
```

But I'm not sure whether I like that some strings are class parts and other strings are something else. Preferably there would be an entire new type with which one could expand theme scales to make it more TypeScript-friendly. A good candidate would be functions, but we use functions for class parts as well already. Is there a good way to distinguish those? 🤔

Other considerations:
- How does this affect bundle size? If we add all the theme objects, this could get quite large.
- How does this affect tailwind-merge initialization? Checking every string in the `classGroups` object could negatively impact performance.

# 2. Helper function to extend theme scales

A simple helper function which converts theme scales to classGroup scales. Could be used like this:

```ts
extendTailwindMerge(themeToClassGroups({
    theme: {
        spacing: ['0', '2xs', 'xs', '2sm', 'sm', …],
        colors: …,
        somethingElse: …
    }
}))
```

This way the bundle size and logic of the rest of tailwind-merge could stay untouched. The problem here is that when using `themeToClassGroups`, the bundle size would probably get even bigger than with option 1. Also removing values from a theme scale would still require a lot of manual work from the library user.

# 3. Separate tool to generate tailwind-merge config

I imagine a website with a text field where you can paste your [resolved Tailwind config](https://tailwindcss.com/docs/configuration#referencing-in-java-script) and you get back the `extendTailwindMerge({ … })` config back. This would have no impact on the tailwind-merge bundle size. The problem here would be that library users are forced to use this tool whenever they change their Tailwind config which could also be annoying.

---

So far I'm in favor of option 1, it feels like the one with the best DX in the long term. But it's also the option which puts the most pressure on the runtime, so potentially biggest impact on UX.

**Comment by dcastil** [comment]
@nenostra I just released [v0.9.0](https://github.com/dcastil/tailwind-merge/releases/tag/v0.9.0) with a `theme` property in the config just like described in the first option of my comment above. Feel free to check it out, it should make your config a lot easier. 😊
